File: module-6-context/solutions/context-agent.ts
/**
 * Module 6: Context-aware coding agent
 * Module 5 agent + compaction + scratchpad + delegation
 */
import "dotenv/config";
import Anthropic from "@anthropic-ai/sdk";
import * as path from "path";
import { Sandbox } from "../../module-4-filesystem/solutions/sandbox.js";
import { createFileTools } from "../../module-4-filesystem/solutions/tools.js";
import { createEditFileTool } from "../../module-5-coding-agent/solutions/edit-file.js";
import { CODING_AGENT_PROMPT } from "../../module-5-coding-agent/solutions/prompt.js";
import { ToolRegistry } from "../../module-3-tools/solutions/tool-registry.js";
import { compactHistory, estimateTokens } from "./compaction.js";
import { createScratchpadTools } from "./scratchpad.js";
import { createDelegateTaskTool } from "./sub-agent.js";
import type { Message } from "../../module-2-agent-loop/solutions/types.js";

const client = new Anthropic();

async function contextAgent(task: string, workspacePath: string): Promise<string> {
  const sandbox = new Sandbox(path.resolve(workspacePath));
  const registry = new ToolRegistry();

  for (const tool of createFileTools(sandbox)) registry.register(tool);
  registry.register(createEditFileTool(sandbox));
  for (const tool of createScratchpadTools(sandbox.root)) registry.register(tool);

  const systemPrompt = CODING_AGENT_PROMPT + `\nWorkspace: ${sandbox.root}\n\nYou also have a scratchpad for persistent notes and can delegate subtasks to sub-agents. Use the scratchpad to track progress on complex tasks. Use delegation for self-contained subtasks.`;

  registry.register(createDelegateTaskTool(registry, systemPrompt));

  let messages: Message[] = [{ role: "user", content: task }];
  const maxTurns = 40;
  const tokenLimit = 80000;

  for (let turn = 0; turn < maxTurns; turn++) {
    // Compact if approaching token limit
    messages = await compactHistory(messages, tokenLimit);

    const tokens = estimateTokens(messages);
    console.log(`\n[Turn ${turn + 1}/${maxTurns}] (~${tokens} tokens, ${messages.length} messages)`);

    const response = await client.messages.create({
      model: "claude-sonnet-4-20250514",
      max_tokens: 4096,
      system: systemPrompt,
      tools: registry.getDefinitions(),
      messages,
    });

    messages.push({ role: "assistant", content: response.content });

    for (const block of response.content) {
      if (block.type === "text") console.log(block.text);
    }

    if (response.stop_reason === "end_turn") {
      const text = response.content.find((b) => b.type === "text");
      return text?.type === "text" ? text.text : "(done)";
    }

    const toolResults: Anthropic.ToolResultBlockParam[] = [];
    for (const block of response.content) {
      if (block.type === "tool_use") {
        const inputStr = JSON.stringify(block.input);
        console.log(`  → ${block.name}(${inputStr.slice(0, 120)}${inputStr.length > 120 ? "..." : ""})`);
        const result = await registry.execute(block.name, block.input as Record<string, unknown>);
        console.log(`  ← ${result.slice(0, 200)}${result.length > 200 ? "..." : ""}`);
        toolResults.push({ type: "tool_result", tool_use_id: block.id, content: result });
      }
    }
    messages.push({ role: "user", content: toolResults });
  }

  return "(max turns reached)";
}

async function main() {
  const task = process.argv[2];
  if (!task) {
    console.error("Usage: npx tsx module-6-context/context-agent.ts \"<task>\" [workspace-path]");
    process.exit(1);
  }
  const workspace = process.argv[3] || process.cwd();
  console.log(`Task: ${task}`);
  console.log(`Workspace: ${workspace}`);
  const result = await contextAgent(task, workspace);
  console.log(`\n=== Result ===\n${result}`);
}

main().catch(console.error);

File: module-6-context/solutions/python/context_agent.py
"""Module 6: Context-aware coding agent."""
import sys
import os
import json

sys.path.insert(0, os.path.dirname(__file__))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-4-filesystem", "solutions", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-5-coding-agent", "solutions", "python"))
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "module-3-tools", "solutions", "python"))

from dotenv import load_dotenv
import anthropic
from sandbox import Sandbox
from tools import create_file_tools
from edit_file import create_edit_file_tool
from prompt import CODING_AGENT_PROMPT
from tool_registry import ToolRegistry, Tool
from compaction import compact_history, estimate_tokens
from scratchpad import create_scratchpad_tools
from sub_agent import create_delegate_task_tool

load_dotenv()

client = anthropic.Anthropic()


def context_agent(task: str, workspace_path: str) -> str:
    sandbox = Sandbox(os.path.realpath(workspace_path))
    registry = ToolRegistry()

    for t in create_file_tools(sandbox):
        registry.register(Tool(**t))
    registry.register(Tool(**create_edit_file_tool(sandbox)))
    for t in create_scratchpad_tools(sandbox.root):
        registry.register(Tool(**t))

    system_prompt = (
        CODING_AGENT_PROMPT
        + f"\nWorkspace: {sandbox.root}\n\n"
        + "You also have a scratchpad for persistent notes and can delegate subtasks to sub-agents. "
        + "Use the scratchpad to track progress on complex tasks. Use delegation for self-contained subtasks."
    )

    delegate_dict = create_delegate_task_tool(registry, system_prompt)
    registry.register(Tool(**delegate_dict))

    messages = [{"role": "user", "content": task}]
    max_turns = 40
    token_limit = 80000

    for turn in range(max_turns):
        messages = compact_history(messages, token_limit)
        tokens = estimate_tokens(messages)
        print(f"\n[Turn {turn + 1}/{max_turns}] (~{tokens} tokens, {len(messages)} messages)")

        response = client.messages.create(
            model="claude-sonnet-4-20250514",
            max_tokens=4096,
            system=system_prompt,
            tools=registry.get_definitions(),
            messages=messages,
        )

        messages.append({"role": "assistant", "content": response.content})

        for block in response.content:
            if block.type == "text":
                print(block.text)

        if response.stop_reason == "end_turn":
            for block in response.content:
                if block.type == "text":
                    return block.text
            return "(done)"

        tool_results = []
        for block in response.content:
            if block.type == "tool_use":
                input_str = json.dumps(block.input)
                print(f"  → {block.name}({input_str[:120]}{'...' if len(input_str) > 120 else ''})")
                result = registry.execute(block.name, block.input)
                print(f"  ← {result[:200]}{'...' if len(result) > 200 else ''}")
                tool_results.append({"type": "tool_result", "tool_use_id": block.id, "content": result})
        messages.append({"role": "user", "content": tool_results})

    return "(max turns reached)"


def main():
    if len(sys.argv) < 2:
        print('Usage: python context_agent.py "<task>" [workspace-path]')
        sys.exit(1)
    task = sys.argv[1]
    workspace = sys.argv[2] if len(sys.argv) > 2 else os.getcwd()
    print(f"Task: {task}")
    print(f"Workspace: {workspace}")
    result = context_agent(task, workspace)
    print(f"\n=== Result ===\n{result}")


if __name__ == "__main__":
    main()
